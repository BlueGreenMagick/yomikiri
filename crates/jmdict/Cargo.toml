[package]
name = "yomikiri-jmdict"
version = "0.1.0"
edition = "2021"
description = "JMDict XML Types"
license = "MIT"

[features]
wasm = ["dep:tsify-next", "dep:wasm-bindgen"]

[dependencies]
thiserror = { workspace = true }
lazy_static = { workspace = true }
quick-xml = { workspace = true }
serde = { workspace = true }
log = { workspace = true }
polonius-the-crab = { workspace = true }
lazy-regex = { workspace = true }

tsify-next = { workspace = true, optional = true }
wasm-bindgen = { workspace = true, optional = true }

[dev-dependencies]
insta = { workspace = true }

[lints]
workspace = true
