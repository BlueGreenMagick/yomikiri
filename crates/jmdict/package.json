{
  "name": "@yomikiri/jmdict-rs",
  "collaborators": [
    "BlueGreenMagick <bluegreenmagick@gmail.com>"
  ],
  "description": "generated JMDict format usable in Yomikiri",
  "version": "0.1.0",
  "license": "MIT",
  "scripts": {}
}