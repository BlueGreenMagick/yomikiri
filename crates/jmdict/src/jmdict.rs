#[derive(Debug, Default, PartialEq, Eq)]
pub struct JMEntry {
    /// 0+ k_ele
    pub forms: Vec<JMForm>,
    /// 1+ r_ele
    pub readings: Vec<JMReading>,
    /// 1+ sense
    pub senses: Vec<JMSense>,
}

impl JMEntry {
    pub fn priority(&self) -> u16 {
        let priorities = &self.readings.first().unwrap().priority;
        let mut priority: u16 = 0;

        for p in priorities {
            // common ~20k entries
            if ["news1", "ichi1", "spec1", "gai1"].contains(&p.as_str()) {
                if priority < 100 {
                    priority += 100
                } else {
                    priority += 25
                }
            // common ~30k entries
            } else if ["news2", "ichi2", "spec2", "gai2"].contains(&p.as_str()) {
                priority += 5
            // 01 ~ 48, each with ~500 entries
            } else if let Some(stripped) = p.strip_prefix("nf") {
                let freq = stripped
                    .parse::<u16>()
                    .expect("could not parse XX as number where priority nfXX");
                priority += 50 - freq;
            }
        }
        priority
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct JMForm {
    pub form: String,
    /// 0+ ke_inf
    pub info: Vec<String>,
    /// 0+ ke_pri
    pub priority: Vec<String>,
}

impl JMForm {
    pub fn is_uncommon(&self) -> bool {
        for f in ["=ok=", "=rK=", "=sK="] {
            if self.info.iter().any(|s| s == f) {
                return true;
            }
        }
        false
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct JMReading {
    pub reading: String,
    pub nokanji: bool,
    /// 0+ re_restr
    pub to_form: Vec<String>,
    /// 0+ re_inf
    pub info: Vec<String>,
    /// 0+ re_pri
    pub priority: Vec<String>,
}

impl JMReading {
    pub fn is_uncommon(&self) -> bool {
        for f in ["=ok=", "=sk="] {
            if self.info.iter().any(|s| s == f) {
                return true;
            }
        }
        false
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct JMSense {
    /// 0+ 'stagk'
    pub to_form: Vec<String>,
    /// 0+ 'stagr'
    pub to_reading: Vec<String>,
    /// 'pos'
    pub part_of_speech: Vec<String>,
    /// 'xref'
    // pub reference: Vec<String>,
    /// 'ant'
    // pub antonym: Vec<String>,
    /// 'field'
    // pub field: Vec<String>,
    /// 'misc'
    pub misc: Vec<String>,
    /// 's_inf'
    pub info: Vec<String>,
    /// 'dial'
    pub dialect: Vec<String>,
    /// 'gloss'
    pub meaning: Vec<String>,
    // 'example'
    // example: Vec<Example>,
}
