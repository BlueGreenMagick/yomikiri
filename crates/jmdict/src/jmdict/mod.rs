mod parse;
mod types;

pub use parse::{parse_jmdict_xml, JMDictParser};
pub use types::*;
