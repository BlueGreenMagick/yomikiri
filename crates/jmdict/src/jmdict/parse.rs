use core::str;
use std::io::BufRead;

use log::warn;
use quick_xml::events::Event;
use quick_xml::Reader;

use super::types::{
    JMDialect, JMDict, JMEntry, JMKanji, JMKanjiInfo, JMPartOfSpeech, JMReading, JMSense,
    JMSenseMisc,
};
use crate::jmdict::types::JMReadingInfo;
use crate::utils::parse_entity_enum_into;
use crate::xml::{get_next_child_in, parse_string_in_tag_into, TagName};
use crate::{Error, Result};

struct JMDictParser<R: BufRead> {
    reader: Reader<R>,
    buf: Vec<u8>,
}

impl<R: BufRead> JMDictParser<R> {
    pub fn new(reader: R) -> Result<Self> {
        let mut reader = Reader::from_reader(reader);
        reader.config_mut().trim_text_start = true;
        reader.config_mut().trim_text_end = true;
        reader.config_mut().expand_empty_elements = true;

        let buf = Vec::new();
        let mut parser = JMDictParser { reader, buf };
        parser.parse_jmdict_start()?;
        Ok(parser)
    }

    fn parse_jmdict_start(&mut self) -> Result<()> {
        loop {
            self.buf.clear();
            match self.reader.read_event_into(&mut self.buf)? {
                Event::Start(tag) => match tag.name().0 {
                    b"JMdict" => return Ok(()),
                    _ => {
                        warn!("Unknown global tag: {}", tag.tag_name());
                    }
                },
                Event::Eof => return Err(Error::InvalidXml("<JMdict> not found".into())),
                _ => {}
            }
        }
    }

    fn parse_entry_in_jmdict(&mut self) -> Result<Option<JMEntry>> {
        loop {
            let start = match get_next_child_in(&mut self.reader, &mut self.buf, "JMdict")? {
                Some(tag) => tag,
                None => return Ok(None),
            };
            match start.name().0 {
                b"entry" => return self.parse_in_entry().map(Some),
                _ => {
                    warn!("Unknown tag in <jmdict>: <{}>", start.tag_name());
                }
            }
        }
    }

    fn parse_in_entry(&mut self) -> Result<JMEntry> {
        let mut id: Option<u32> = None;
        let mut entry = JMEntry::default();

        while let Some(start) = get_next_child_in(&mut self.reader, &mut self.buf, "entry")? {
            match start.name().0 {
                b"ent_seq" => {
                    if let Some(id) = id {
                        return Err(Error::MultipleEntryIds(id));
                    }
                    let idstr =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"ent_seq")?;
                    let seq_id = str::parse::<u32>(&idstr)
                        .map_err(|_| format!("Couldn't parse as u32 number: {}", idstr))?;
                    id = Some(seq_id);
                }
                b"k_ele" => {
                    let form = self.parse_in_form()?;
                    entry.kanjis.push(form);
                }
                b"r_ele" => {
                    let reading = self.parse_in_reading()?;
                    entry.readings.push(reading);
                }
                b"sense" => {
                    let sense = self.parse_in_sense()?;
                    entry.senses.push(sense);
                }
                _ => {
                    warn!("Unknown tag in <entry>: <{}>", start.tag_name())
                }
            };
        }

        if let Some(id) = id {
            entry.id = id;
            return Ok(entry);
        } else {
            return Err(Error::NoEntryId(self.reader.buffer_position()));
        }
    }

    fn parse_in_form(&mut self) -> Result<JMKanji> {
        let mut form = JMKanji::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "k_ele")? {
            match tag.name().0 {
                b"keb" => {
                    if !form.kanji.is_empty() {
                        warn!("Found multiple <keb> in form '{}' in JMDict", form.kanji)
                    }
                    form.kanji = parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"keb")?;
                }
                b"ke_inf" => {
                    parse_entity_enum_into!(
                        JMKanjiInfo,
                        &mut self.reader,
                        &mut self.buf,
                        "ke_inf",
                        form.info
                    );
                }
                b"ke_pri" => {
                    form.priority.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"ke_pri",
                    )?);
                }
                _ => {
                    warn!("Unknown tag in <entry>: {}", tag.tag_name());
                }
            };
        }

        Ok(form)
    }

    fn parse_in_reading(&mut self) -> Result<JMReading> {
        let mut reading = JMReading::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "r_ele")? {
            match tag.name().0 {
                b"reb" => {
                    reading.reading =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"reb")?;
                }
                b"re_nokanji" => {
                    reading.nokanji = true;
                    // consume ending tag
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"re_nokanji")?;
                }
                b"re_restr" => {
                    reading.to_form.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"re_restr",
                    )?);
                }
                b"re_inf" => {
                    parse_entity_enum_into!(
                        JMReadingInfo,
                        &mut self.reader,
                        &mut self.buf,
                        "re_inf",
                        reading.info
                    );
                }
                b"re_pri" => {
                    reading.priority.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"re_pri",
                    )?);
                }
                _ => {
                    warn!("Unknown tag in <r_ele>: {}", &tag.tag_name());
                }
            };
        }

        Ok(reading)
    }

    fn parse_in_sense(&mut self) -> Result<JMSense> {
        let mut sense = JMSense::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "sense")? {
            match tag.name().0 {
                b"stagk" => {
                    sense.to_form.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"stagk",
                    )?);
                }
                b"stagr" => {
                    sense.to_reading.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"stagr",
                    )?);
                }
                b"pos" => {
                    parse_entity_enum_into!(
                        JMPartOfSpeech,
                        &mut self.reader,
                        &mut self.buf,
                        "pos",
                        sense.pos
                    );
                }
                b"xref" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"xref")?;
                }
                b"ant" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"ant")?;
                }
                b"field" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"field")?;
                }
                b"misc" => {
                    parse_entity_enum_into!(
                        JMSenseMisc,
                        &mut self.reader,
                        &mut self.buf,
                        "misc",
                        sense.misc
                    );
                }
                b"s_inf" => {
                    sense.info.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"s_inf",
                    )?);
                }
                b"dial" => {
                    parse_entity_enum_into!(
                        JMDialect,
                        &mut self.reader,
                        &mut self.buf,
                        "dial",
                        sense.dialects
                    );
                }
                b"gloss" => {
                    sense.meanings.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"gloss",
                    )?);
                }
                b"lsource" => {
                    // consume ending tag
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"lsource")?;
                }
                b"example" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"example")?;
                }
                _ => {
                    warn!("Unknown tag in <sense>: {}", tag.tag_name());
                }
            };
        }

        Ok(sense)
    }
}

pub fn parse_jmdict_xml(xml: &str) -> Result<JMDict> {
    let mut parser = JMDictParser::new(xml.as_bytes())?;
    let mut entries = vec![];
    while let Some(entry) = parser.parse_entry_in_jmdict()? {
        entries.push(entry);
    }
    Ok(JMDict { entries })
}
