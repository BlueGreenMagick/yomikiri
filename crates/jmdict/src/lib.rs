pub mod error;
pub mod jmdict;
pub mod jmnedict;
mod utils;
mod xml;

pub use crate::error::Error;
pub use jmdict::parse_jmdict_xml;

pub type Result<T> = core::result::Result<T, Error>;
