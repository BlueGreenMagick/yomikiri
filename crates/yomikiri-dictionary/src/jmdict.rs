use crate::entry::{Entry, Form, PartOfSpeech, Reading, Sense};
use crate::Result;
use itertools::Itertools;
use yomikiri_jmdict::jmdict::{JMEntry, JMForm, JMReading, JMSense};

pub fn parse_jmdict_xml(xml: &str) -> Result<Vec<Entry>> {
    let jmdict = yomikiri_jmdict::parse_jmdict_xml(xml)?;
    let entries = jmdict.entries.into_iter().map(Entry::from).collect();
    Ok(entries)
}

impl From<JMEntry> for Entry {
    fn from(jm_entry: JMEntry) -> Entry {
        let priority = jm_entry.priority();
        let forms: Vec<Form> = jm_entry.forms.into_iter().map(Form::from).collect();
        let readings: Vec<Reading> = jm_entry.readings.into_iter().map(Reading::from).collect();
        let senses: Vec<Sense> = jm_entry.senses.into_iter().map(Sense::from).collect();
        Entry {
            forms,
            readings,
            senses,
            priority,
        }
    }
}

impl From<JMForm> for Form {
    fn from(jm_form: JMForm) -> Self {
        let uncommon = jm_form.is_uncommon();
        Form {
            form: jm_form.form,
            info: jm_form.info,
            uncommon,
        }
    }
}

impl From<JMReading> for Reading {
    fn from(jm_reading: JMReading) -> Self {
        let uncommon = jm_reading.is_uncommon();
        Reading {
            reading: jm_reading.reading,
            nokanji: jm_reading.nokanji,
            to_form: jm_reading.to_form,
            info: jm_reading.info,
            uncommon,
        }
    }
}

impl From<JMSense> for Sense {
    fn from(jm_sense: JMSense) -> Self {
        Sense {
            to_form: jm_sense.to_form,
            to_reading: jm_sense.to_reading,
            pos: jm_sense
                .part_of_speech
                .iter()
                .map(|s| parse_part_of_speech(s))
                .unique()
                .collect(),
            misc: jm_sense.misc,
            info: jm_sense.info,
            dialect: jm_sense.dialect,
            meaning: jm_sense.meaning,
        }
    }
}

fn parse_part_of_speech(value: &str) -> PartOfSpeech {
    // strip '=' from '=XXX='
    let value = &value[1..value.len() - 1];
    match value {
        "n" | "adj-no" | "adj-f" | "num" | "vs" => PartOfSpeech::Noun,
        "adv" | "adv-to" | "vs-c" | "vs-i" => PartOfSpeech::Adverb,
        "conj" => PartOfSpeech::Conjunction,
        "int" => PartOfSpeech::Interjection,
        "suf" | "n-suf" | "ctr" => PartOfSpeech::Suffix,
        "prt" => PartOfSpeech::Particle,
        "adj-i" | "adj-ix" => PartOfSpeech::Adjective,
        "adj-na" | "adj-t" | "adj-nari" => PartOfSpeech::NaAdjective,
        "aux-v" | "aux" | "aux-adj" | "cop" => PartOfSpeech::AuxiliaryVerb,
        "pn" => PartOfSpeech::Pronoun,
        "pref" => PartOfSpeech::Prefix,
        "adj-pn" => PartOfSpeech::Adnomial,
        "exp" => PartOfSpeech::Expression,
        "unc" => PartOfSpeech::Unclassified,
        s if s.starts_with('v') => PartOfSpeech::Verb,
        s if s.starts_with("adj-") => PartOfSpeech::Adjective,
        s if s.starts_with("n-") => PartOfSpeech::Noun,
        other => panic!("Unknown part of speech: {}", other),
    }
}
