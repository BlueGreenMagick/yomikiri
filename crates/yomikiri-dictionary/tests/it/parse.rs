use yomikiri_dictionary::jmdict::parse_jmdict_xml;

type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

#[test]
fn test_jmdict_parse_1() -> Result<()> {
    let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<!-- JMdict created: 2024-08-23 -->
<JMdict>
<entry>
<ent_seq>1000810</ent_seq>
<k_ele>
<keb>いじり回す</keb>
</k_ele>
<k_ele>
<keb>弄り回す</keb>
</k_ele>
<k_ele>
<keb>弄りまわす</keb>
</k_ele>
<r_ele>
<reb>いじりまわす</reb>
</r_ele>
<sense>
<pos>&v5s;</pos>
<pos>&vt;</pos>
<gloss>to tinker with</gloss>
<gloss>to fumble with</gloss>
<gloss>to fiddle with</gloss>
<gloss>to twiddle</gloss>
</sense>
</entry>
</JMdict>
"#;
    let entries = parse_jmdict_xml(&xml)?;
    insta::assert_yaml_snapshot!(entries);
    Ok(())
}

#[test]
fn test_jmdict_parse_2() -> Result<()> {
    let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<!-- JMdict created: 2024-08-23 -->
<JMdict>
<entry>
<ent_seq>1006950</ent_seq>
<k_ele>
<keb>抑</keb>
<ke_inf>&rK;</ke_inf>
</k_ele>
<k_ele>
<keb>抑々</keb>
<ke_inf>&rK;</ke_inf>
</k_ele>
<k_ele>
<keb>抑抑</keb>
<ke_inf>&rK;</ke_inf>
</k_ele>
<k_ele>
<keb>抑も</keb>
<ke_inf>&io;</ke_inf>
<ke_inf>&rK;</ke_inf>
</k_ele>
<r_ele>
<reb>そもそも</reb>
</r_ele>
<sense>
<pos>&n;</pos>
<pos>&adv;</pos>
<misc>&uk;</misc>
<gloss>in the first place</gloss>
</sense>
<sense>
<pos>&conj;</pos>
<misc>&uk;</misc>
<s_inf>used when bringing up something already mentioned</s_inf>
<gloss>after all</gloss>
<gloss>anyway</gloss>
</sense>
<sense>
<pos>&n;</pos>
<pos>&adj-no;</pos>
<misc>&uk;</misc>
<gloss>beginning</gloss>
<gloss>start</gloss>
</sense>
</entry>
</JMdict>
"#;
    let entries = parse_jmdict_xml(&xml)?;
    insta::assert_yaml_snapshot!(entries);
    Ok(())
}
