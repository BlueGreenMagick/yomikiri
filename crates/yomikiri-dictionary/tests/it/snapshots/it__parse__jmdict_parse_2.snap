---
source: crates/yomikiri-dictionary/tests/it/parse.rs
assertion_line: 91
expression: entries
---
- id: 1006950
  kanjis:
    - kanji: 抑
      rarity: rare
    - kanji: 抑々
      rarity: rare
    - kanji: 抑抑
      rarity: rare
    - kanji: 抑も
      rarity: rare
  readings:
    - reading: そもそも
      nokanji: false
      toKanji: []
      rarity: normal
  groupedSenses:
    - pos:
        - noun
        - adverb
      senses:
        - meanings:
            - in the first place
          toKanji: []
          toReading: []
          misc:
            - usuallyKanaAlone
          info: []
          dialects: []
    - pos:
        - conjunction
      senses:
        - meanings:
            - after all
            - anyway
          toKanji: []
          toReading: []
          misc:
            - usuallyKanaAlone
          info:
            - used when bringing up something already mentioned
          dialects: []
    - pos:
        - noun
      senses:
        - meanings:
            - beginning
            - start
          toKanji: []
          toReading: []
          misc:
            - usuallyKanaAlone
          info: []
          dialects: []
  priority: 0
